File: src/simulation/physics/fmm/octree.js
/**
 * Octree для FMM: одна и та же структура узлов с мультиполями и листовыми индексами.
 * Построение по источникам (particles) с полями x,y,z, gamma, vorticity, coreRadius.
 */

function getOctant(source, cx, cy, cz) {
  let octant = 0
  if (source.x >= cx) octant |= 1
  if (source.y >= cy) octant |= 2
  if (source.z >= cz) octant |= 4
  return octant
}

export function computeBounds(sources) {
  let minX = Infinity
  let minY = Infinity
  let minZ = Infinity
  let maxX = -Infinity
  let maxY = -Infinity
  let maxZ = -Infinity

  for (let i = 0; i < sources.length; i += 1) {
    const s = sources[i]
    if (s.x < minX) minX = s.x
    if (s.y < minY) minY = s.y
    if (s.z < minZ) minZ = s.z
    if (s.x > maxX) maxX = s.x
    if (s.y > maxY) maxY = s.y
    if (s.z > maxZ) maxZ = s.z
  }

  const centerX = (minX + maxX) * 0.5
  const centerY = (minY + maxY) * 0.5
  const centerZ = (minZ + maxZ) * 0.5
  const extent = Math.max(maxX - minX, maxY - minY, maxZ - minZ, 1e-4)
  return {
    centerX,
    centerY,
    centerZ,
    halfSize: extent * 0.5 + 1e-4,
  }
}

/**
 * Строит узел дерева. sources — массив { x,y,z, omegaGammaX,Y,Z, sigma, id }, indices — индексы в sources.
 * Возвращает узел с полями: leaf, indices, children, centerX,Y,Z, halfSize, count, multipole { comX, comY, comZ, omegaGammaX,Y,Z, sigmaMean }.
 */
export function buildNode(sources, indices, cx, cy, cz, halfSize, leafSize, depth, maxDepth) {
  if (indices.length === 0) {
    return null
  }

  let totalWeight = 0
  let weightedX = 0
  let weightedY = 0
  let weightedZ = 0
  let sigmaWeighted = 0
  let omegaGammaX = 0
  let omegaGammaY = 0
  let omegaGammaZ = 0

  for (let i = 0; i < indices.length; i += 1) {
    const s = sources[indices[i]]
    const w = Math.abs(s.gamma ?? 1) + 1e-6
    totalWeight += w
    weightedX += s.x * w
    weightedY += s.y * w
    weightedZ += s.z * w
    sigmaWeighted += (s.sigma ?? 0.01) * w
    omegaGammaX += s.omegaGammaX ?? 0
    omegaGammaY += s.omegaGammaY ?? 0
    omegaGammaZ += s.omegaGammaZ ?? 0
  }

  const node = {
    leaf: indices.length <= leafSize || depth >= maxDepth,
    indices: null,
    children: null,
    centerX: cx,
    centerY: cy,
    centerZ: cz,
    halfSize,
    count: indices.length,
    multipole: (() => {
      const comXv = weightedX / Math.max(totalWeight, 1e-6)
      const comYv = weightedY / Math.max(totalWeight, 1e-6)
      const comZv = weightedZ / Math.max(totalWeight, 1e-6)
      let qTrace = 0
      for (let qi = 0; qi < indices.length; qi += 1) {
        const qs = sources[indices[qi]]
        const dx = qs.x - comXv
        const dy = qs.y - comYv
        const dz = qs.z - comZv
        const strength = Math.hypot(qs.omegaGammaX ?? 0, qs.omegaGammaY ?? 0, qs.omegaGammaZ ?? 0)
        qTrace += strength * (dx * dx + dy * dy + dz * dz)
      }
      return {
        comX: comXv, comY: comYv, comZ: comZv,
        omegaGammaX, omegaGammaY, omegaGammaZ,
        sigmaMean: Math.max(1e-6, sigmaWeighted / Math.max(totalWeight, 1e-6)),
        qTrace,
      }
    })(),
  }

  if (node.leaf) {
    node.indices = indices
    return node
  }

  const childHalf = halfSize * 0.5
  const buckets = Array.from({ length: 8 }, () => [])
  for (let i = 0; i < indices.length; i += 1) {
    const idx = indices[i]
    const s = sources[idx]
    buckets[getOctant(s, cx, cy, cz)].push(idx)
  }

  node.children = new Array(8).fill(null)
  for (let octant = 0; octant < 8; octant += 1) {
    const childIndices = buckets[octant]
    if (childIndices.length === 0) continue

    const childCx = cx + ((octant & 1) !== 0 ? childHalf : -childHalf)
    const childCy = cy + ((octant & 2) !== 0 ? childHalf : -childHalf)
    const childCz = cz + ((octant & 4) !== 0 ? childHalf : -childHalf)
    node.children[octant] = buildNode(
      sources,
      childIndices,
      childCx,
      childCy,
      childCz,
      childHalf,
      leafSize,
      depth + 1,
      maxDepth,
    )
  }

  return node
}

/**
 * Собирает все листовые узлы в массив (для построения interaction lists и обхода).
 */
export function collectLeaves(node, out = []) {
  if (!node) return out
  if (node.leaf) {
    out.push(node)
    return out
  }
  for (let i = 0; i < node.children.length; i += 1) {
    collectLeaves(node.children[i], out)
  }
  return out
}

/**
 * Проверка well-separated: (sizeA + sizeB) <= theta * distance(centerA, centerB).
 */
export function areWellSeparated(leafA, leafB, theta) {
  const sizeA = leafA.halfSize * 2
  const sizeB = leafB.halfSize * 2
  const dx = leafA.centerX - leafB.centerX
  const dy = leafA.centerY - leafB.centerY
  const dz = leafA.centerZ - leafB.centerZ
  const dist = Math.sqrt(dx * dx + dy * dy + dz * dz) + 1e-12
  return sizeA + sizeB <= theta * dist
}

File: torus-core/crates/torus-fmm/src/octree.rs
//! Octree for FMM: adaptive spatial decomposition.
//!
//! Port of `src/simulation/physics/fmm/octree.js`.

use glam::DVec3;

use crate::multipole::{FmmSource, Multipole, merge_multipoles};

/// Octree node.
pub struct OctreeNode {
    pub leaf: bool,
    pub center: DVec3,
    pub half_size: f64,
    pub count: usize,
    pub multipole: Multipole,
    /// Leaf-only: indices into source array.
    pub indices: Vec<usize>,
    /// Internal-only: 8 children (octants).
    pub children: [Option<Box<OctreeNode>>; 8],
}

/// Bounding box of sources.
pub struct Bounds {
    pub center: DVec3,
    pub half_size: f64,
}

pub fn compute_bounds(sources: &[FmmSource]) -> Bounds {
    if sources.is_empty() {
        return Bounds {
            center: DVec3::ZERO,
            half_size: 1.0,
        };
    }

    let mut min = DVec3::splat(f64::INFINITY);
    let mut max = DVec3::splat(f64::NEG_INFINITY);

    for s in sources {
        min = min.min(s.position);
        max = max.max(s.position);
    }

    let center = (min + max) * 0.5;
    let extent = (max - min).max_element().max(1e-4);

    Bounds {
        center,
        half_size: extent * 0.5 + 1e-4,
    }
}

fn get_octant(pos: DVec3, center: DVec3) -> usize {
    let mut octant = 0;
    if pos.x >= center.x { octant |= 1; }
    if pos.y >= center.y { octant |= 2; }
    if pos.z >= center.z { octant |= 4; }
    octant
}

fn child_center(parent_center: DVec3, child_half: f64, octant: usize) -> DVec3 {
    DVec3::new(
        parent_center.x + if octant & 1 != 0 { child_half } else { -child_half },
        parent_center.y + if octant & 2 != 0 { child_half } else { -child_half },
        parent_center.z + if octant & 4 != 0 { child_half } else { -child_half },
    )
}

/// Build an octree node recursively.
pub fn build_node(
    sources: &[FmmSource],
    indices: &[usize],
    center: DVec3,
    half_size: f64,
    leaf_size: usize,
    depth: u32,
    max_depth: u32,
) -> Option<Box<OctreeNode>> {
    if indices.is_empty() {
        return None;
    }

    // Compute multipole for this node
    let mut total_weight = 0.0;
    let mut weighted_pos = DVec3::ZERO;
    let mut sigma_weighted = 0.0;
    let mut omega_gamma = DVec3::ZERO;

    for &idx in indices {
        let s = &sources[idx];
        let w = s.gamma.abs() + 1e-6;
        total_weight += w;
        weighted_pos += s.position * w;
        sigma_weighted += s.sigma * w;
        omega_gamma += s.omega_gamma;
    }

    let com = weighted_pos / total_weight.max(1e-6);
    let sigma_mean = (sigma_weighted / total_weight.max(1e-6)).max(1e-6);

    let mut q_trace = 0.0;
    for &idx in indices {
        let s = &sources[idx];
        let d = s.position - com;
        let strength = s.omega_gamma.length();
        q_trace += strength * d.length_squared();
    }

    let multipole = Multipole {
        com,
        omega_gamma,
        sigma_mean,
        q_trace,
    };

    let is_leaf = indices.len() <= leaf_size || depth >= max_depth;

    if is_leaf {
        return Some(Box::new(OctreeNode {
            leaf: true,
            center,
            half_size,
            count: indices.len(),
            multipole,
            indices: indices.to_vec(),
            children: Default::default(),
        }));
    }

    // Split into 8 octants
    let child_half = half_size * 0.5;
    let mut buckets: [Vec<usize>; 8] = Default::default();

    for &idx in indices {
        let octant = get_octant(sources[idx].position, center);
        buckets[octant].push(idx);
    }

    let mut children: [Option<Box<OctreeNode>>; 8] = Default::default();
    for octant in 0..8 {
        if buckets[octant].is_empty() {
            continue;
        }
        let cc = child_center(center, child_half, octant);
        children[octant] = build_node(
            sources,
            &buckets[octant],
            cc,
            child_half,
            leaf_size,
            depth + 1,
            max_depth,
        );
    }

    Some(Box::new(OctreeNode {
        leaf: false,
        center,
        half_size,
        count: indices.len(),
        multipole,
        indices: Vec::new(),
        children,
    }))
}

/// Upward pass: recompute multipoles for internal nodes from children.
pub fn upward_pass(node: &mut OctreeNode) {
    if node.leaf {
        return;
    }

    let mut child_multipoles: Vec<&Multipole> = Vec::new();
    for child in &mut node.children {
        if let Some(c) = child {
            upward_pass(c);
            child_multipoles.push(&c.multipole);
        }
    }

    if !child_multipoles.is_empty() {
        node.multipole = merge_multipoles(&child_multipoles);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_sources(n: usize) -> Vec<FmmSource> {
        (0..n)
            .map(|i| {
                let t = 2.0 * std::f64::consts::PI * i as f64 / n as f64;
                FmmSource {
                    id: i as u32,
                    index: i,
                    position: DVec3::new(t.cos(), t.sin(), 0.0),
                    gamma: 1.0,
                    sigma: 0.1,
                    omega_gamma: DVec3::new(0.0, 0.0, 0.1),
                }
            })
            .collect()
    }

    #[test]
    fn builds_tree_without_panic() {
        let sources = make_sources(64);
        let bounds = compute_bounds(&sources);
        let indices: Vec<usize> = (0..sources.len()).collect();
        let root = build_node(&sources, &indices, bounds.center, bounds.half_size, 8, 0, 20);
        assert!(root.is_some());
        let root = root.unwrap();
        assert_eq!(root.count, 64);
        assert!(!root.leaf);
    }

    #[test]
    fn small_set_makes_leaf() {
        let sources = make_sources(4);
        let bounds = compute_bounds(&sources);
        let indices: Vec<usize> = (0..sources.len()).collect();
        let root = build_node(&sources, &indices, bounds.center, bounds.half_size, 8, 0, 20);
        assert!(root.is_some());
        assert!(root.unwrap().leaf);
    }

    #[test]
    fn upward_pass_recomputes_multipole() {
        let sources = make_sources(32);
        let bounds = compute_bounds(&sources);
        let indices: Vec<usize> = (0..sources.len()).collect();
        let mut root = build_node(&sources, &indices, bounds.center, bounds.half_size, 4, 0, 20).unwrap();

        let og_before = root.multipole.omega_gamma;
        upward_pass(&mut root);
        let og_after = root.multipole.omega_gamma;

        // Total omega_gamma should be same (it's summed)
        assert!((og_before.z - og_after.z).abs() < 1e-6);
    }
}
